hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateMap.java

```diff
@@ -374,7 +374,12 @@ Pipeline updatePipelineState(PipelineID pipelineID, PipelineState state)
     PipelineQuery query = new PipelineQuery(pipeline);
     if (updatedPipeline.getPipelineState() == PipelineState.OPEN) {
       // for transition to OPEN state add pipeline to query2OpenPipelines
-      query2OpenPipelines.get(query).add(updatedPipeline);
+      List<Pipeline> pipelineList = query2OpenPipelines.get(query);
+      if (pipelineList == null) {
+        pipelineList = new CopyOnWriteArrayList<>();
+        query2OpenPipelines.put(query, pipelineList);
+      }
+      pipelineList.add(updatedPipeline);
```

Comment:
Besides above change, Can you also do the empty check for another condition branch?
    } else {
      // for transition from OPEN to CLOSED state remove pipeline from
      // query2OpenPipelines
      query2OpenPipelines.get(query).remove(pipeline);   <---
    }
There is also a chance that remove operation can throw NPE.